### ISSUE TITLE
[DTensor] RNG tracker does not advance state for CPU tensors on CUDA mesh, causing trunc_normal_ infinite loop

### ISSUE THREAD

**Issue by weifengpy** [pytorch/pytorch#180088]
### 🐛 Describe the bug

Bug

  DTensor's RNG tracker does not advance CPU RNG state when a DTensor has a CUDA mesh but its local tensor is on CPU. This causes nn.init.trunc_normal_ to loop forever since its rejection sampling (introduced in #174997)
  depends on normal_() producing different values each call.
                                                                                                                                                                                                                                   
  Root cause                                                                                                                                                                                                                     

  The DTensor RNG tracker advances the CUDA RNG but the local tensor is on CPU, so values come from the stale CPU RNG:                                                                                                             
   
  CUDA RNG advances: True (1st), True (2nd)                                                                                                                                                                                        
  CPU  RNG advances: False (1st), False (2nd)                                                                                                                                                                                      
                                                                                                                                                                                                                                   
  How to get a CPU DTensor on a CUDA mesh                                                                                                                                                                                          
                                                                                                                                                                                                                                   
  This happens in FSDP seed checkpoint creation:                                                                                                                                                                                   
  ```                                                                                                                                                                                                                               
  with torch.device("meta"):
      model = model_config.build()                                                                                                                                                                                                 
  fully_shard(model, mesh=cuda_mesh)   # DTensors with CUDA mesh
  model.to_empty(device="cpu")          # local tensors now on CPU                                                                                                                                                                 
  model.init_weights()                  # trunc_normal_ → infinite loop       
```                                                                                                                                                     
                                                                                                                                                                                                                                   
  Repro                                                                                                                                                                                                                            
```                                                                                                                                                                                                         
  # torchrun --nproc_per_node=1 repro.py                                                                                                                                                                                         
  import torch, torch.distributed as dist                                                                                                                                                                                          
  from torch.distributed.device_mesh import init_device_mesh
  from torch.distributed.tensor import DTensor, Shard                                                                                                                                                                              
                                                                                                                                                                                                                                 
  dist.init_process_group("nccl")                                                                                                                                                                                                  
  torch.cuda.set_device(0)                                                                                                                                                                                                       
                                                                                                                                                                                                                                   
  mesh = init_device_mesh("cuda", (1,))                                                                                                                                                                                            
  with torch.device("meta"):
      t = torch.empty(4)                                                                                                                                                                                                           
  dt = DTensor.from_local(t, mesh, [Shard(0)])
  dt = torch.empty_like(dt, device="cpu")

  torch.manual_seed(42)
  cpu_before = torch.random.get_rng_state()
  cuda_before = torch.cuda.get_rng_state()
  dt.normal_(0, 1)
  cpu_after = torch.random.get_rng_state()
  cuda_after = torch.cuda.get_rng_state()

  print(f"CUDA RNG advances: {not torch.equal(cuda_before, cuda_after)}")
  print(f"CPU  RNG advances: {not torch.equal(cpu_before, cpu_after)}")

  a = dt.normal_(0, 1).to_local().clone()
  b = torch.empty_like(dt).normal_(0, 1).to_local()
  assert not torch.equal(a, b), f"BUG: RNG not advancing, a == b == {a}"

  dist.destroy_process_group()
```

### Versions

pytorch nightly

cc @ezyang @gchanan @kadeng @msaroufim @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @dcci @aditvenk @xmfan

**Comment by weifengpy** [comment]
we have many aspects of fixing, 
from DTensor side, https://github.com/pytorch/pytorch/pull/180087
from torch/nn side, cc @albanD about the infinite loop in nn/init.py from https://github.com/pytorch/pytorch/pull/174997

cc @tianyu-l : this feels big for prior-Feb pytorch, before https://github.com/pytorch/pytorch/pull/180087, nn.init(DTenosr) just generate same numbers without advancing correct RNG. with 180087, at lease it's stuck in in the while loop
